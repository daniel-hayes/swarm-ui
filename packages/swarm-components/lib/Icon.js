"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.getIconShape = exports.MEDIA_SIZES = exports.ICON_CIRCLED_CLASS = exports.SVG_THIN_STYLE = exports.ICON_CLASS = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _shapeConstants = require("swarm-icons/dist/js/shapeConstants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ICON_CLASS = 'svg';
exports.ICON_CLASS = ICON_CLASS;
var SVG_THIN_STYLE = '--small';
exports.SVG_THIN_STYLE = SVG_THIN_STYLE;
var ICON_CIRCLED_CLASS = 'svg--circled';
exports.ICON_CIRCLED_CLASS = ICON_CIRCLED_CLASS;
var MEDIA_SIZES = {
  xxs: '12',
  xs: '16',
  s: '24',
  m: '36',
  l: '48',
  xl: '72',
  xxl: '120'
};
exports.MEDIA_SIZES = MEDIA_SIZES;

var SMALL_ICON_VARIANT_WHITELIST = _shapeConstants.VALID_SHAPES.filter(function (s) {
  return !s.startsWith('external') && !s.startsWith('meetup');
} // no third party icons // logos use same path for `xs`
);
/**
 * @param {String} shape - icon shape
 * @param {String} size - icon size
 * @returns {String} icon name (with or without suffix)
 */


var getIconShape = function getIconShape(shape, size) {
  if (!SMALL_ICON_VARIANT_WHITELIST.includes(shape)) {
    return shape;
  }

  var suffix = size === 'xxs' || size === 'xs' || size === 's' ? SVG_THIN_STYLE : '';
  return "".concat(shape).concat(suffix);
};

exports.getIconShape = getIconShape;

/**
 * Icon component used to insert an svg icon into a component or page
 *
 * **Accessibility** If an Icon is used on its own without supporting
 * text to explain what it is/does, be a good citizen and pass in an
 * `aria-label` attribute describing what the icon represents
 *
 * @module Icon
 */
var Icon = function Icon(props) {
  var className = props.className,
      shape = props.shape,
      size = props.size,
      color = props.color,
      style = props.style,
      circle = props.circle,
      other = _objectWithoutProperties(props, ["className", "shape", "size", "color", "style", "circle"]);

  var classNames = (0, _classnames["default"])(ICON_CLASS, "svg--".concat(shape), _defineProperty({}, ICON_CIRCLED_CLASS, circle), className, 'svg-icon', 'valign--middle');
  var sizeVal = MEDIA_SIZES[size];
  var allStyles = style || {};

  if (color) {
    allStyles.fill = color;
  }

  return _react["default"].createElement("svg", _extends({
    preserveAspectRatio: "xMinYMin meet",
    width: sizeVal,
    height: sizeVal,
    viewBox: "0 0 ".concat(sizeVal, " ").concat(sizeVal),
    className: classNames,
    role: "img",
    style: allStyles
  }, other), _react["default"].createElement("use", {
    xlinkHref: "#icon-".concat(getIconShape(shape, size))
  }));
};

Icon.defaultProps = {
  size: 's'
};
var _default = Icon;
exports["default"] = _default;
Icon.__docgenInfo = {
  "description": "Icon component used to insert an svg icon into a component or page\n\n**Accessibility** If an Icon is used on its own without supporting\ntext to explain what it is/does, be a good citizen and pass in an\n`aria-label` attribute describing what the icon represents\n\n@module Icon",
  "methods": [],
  "displayName": "Icon",
  "props": {
    "size": {
      "defaultValue": {
        "value": "'s'",
        "computed": false
      },
      "required": false,
      "flowType": {
        "name": "string"
      },
      "description": "Which of our media sizes to render the icon at"
    },
    "shape": {
      "required": true,
      "flowType": {
        "name": "string"
      },
      "description": "The name of the icon shape to render"
    },
    "circle": {
      "required": false,
      "flowType": {
        "name": "boolean"
      },
      "description": "Whether the icon is outlined with a circle"
    },
    "color": {
      "required": false,
      "flowType": {
        "name": "string"
      },
      "description": "What color the icon should be filled with"
    },
    "className": {
      "required": false,
      "flowType": {
        "name": "string"
      },
      "description": "Class applied to svg tag"
    },
    "style": {
      "required": false,
      "flowType": {
        "name": "signature",
        "type": "object",
        "raw": "{}",
        "signature": {
          "properties": []
        }
      },
      "description": "Object of css styles"
    }
  }
};