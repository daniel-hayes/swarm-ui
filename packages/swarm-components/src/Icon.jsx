import React from 'react';
import cx from 'classnames';
import { VALID_SHAPES } from 'swarm-icons/dist/js/shapeConstants';

export const ICON_CLASS = 'svg';
export const SVG_THIN_STYLE = '--small';
export const ICON_CIRCLED_CLASS = 'svg--circled';

export const MEDIA_SIZES = {
	xxs: '12',
	xs: '16',
	s: '24',
	m: '36',
	l: '48',
	xl: '72',
	xxl: '120',
};

const SMALL_ICON_VARIANT_WHITELIST = VALID_SHAPES.filter(
	s => !s.startsWith('external') && !s.startsWith('meetup') // no third party icons // logos use same path for `xs`
);

/**
 * @param {String} shape - icon shape
 * @param {String} size - icon size
 * @returns {String} icon name (with or without suffix)
 */
export const getIconShape = (shape, size) => {
	if (!SMALL_ICON_VARIANT_WHITELIST.includes(shape)) {
		return shape;
	}

	const suffix =
		size === 'xxs' || size === 'xs' || size === 's' ? SVG_THIN_STYLE : '';
	return `${shape}${suffix}`;
};
type Props = {
	/** Which of our media sizes to render the icon at */
	size?: string,
	/** The name of the icon shape to render */
	shape: string,
	/** Whether the icon is outlined with a circle */
	circle?: boolean,
	/** What color the icon should be filled with */
	color?: string,
};

/**
 * Icon component used to insert an svg icon into a component or page
 *
 * **Accessibility** If an Icon is used on its own without supporting
 * text to explain what it is/does, be a good citizen and pass in an
 * `aria-label` attribute describing what the icon represents
 *
 * @module Icon
 */
const Icon = (props: Props) => {
	const { className, shape, size, color, style, circled, ...other } = props;

	const classNames = cx(
		ICON_CLASS,
		`svg--${shape}`,
		{ [ICON_CIRCLED_CLASS]: circled },
		className
	);

	const sizeVal = MEDIA_SIZES[size];

	const allStyles = style || {};
	if (color) {
		allStyles.fill = color;
	}

	return (
		<svg
			preserveAspectRatio="xMinYMin meet"
			width={sizeVal}
			height={sizeVal}
			viewBox={`0 0 ${sizeVal} ${sizeVal}`}
			className="svg-icon valign--middle"
			role="img"
			style={allStyles}
			{...other}
		>
			<use xlinkHref={`#icon-${getIconShape(shape, size)}`} />
		</svg>
	);
};

Icon.defaultProps = {
	size: 's',
};

export default Icon;
