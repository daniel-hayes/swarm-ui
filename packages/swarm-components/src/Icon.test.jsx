import React from 'react';
import { shallow } from 'enzyme';
import Icon, { MEDIA_SIZES } from './Icon';

describe('Icon', () => {
	const getIconMarkup = (key, props) => (
		<div key={key}>
			{Object.keys(MEDIA_SIZES).map(size => (
				<Icon key={size} size={size} {...props} />
			))}
		</div>
	);

	const testCases = [
		['Regular', getIconMarkup("1", {shape: "cog"})],
		['Circled', getIconMarkup("2", {circle: true, shape: "cog"})],
		['Circled color', getIconMarkup("3", {color: "blue", circle: true, shape: "cog"})],
	];

	test.each(testCases)('Snapshot: %s', (description, element) => {
		expect(shallow(element)).toMatchSnapshot();
	});
});
