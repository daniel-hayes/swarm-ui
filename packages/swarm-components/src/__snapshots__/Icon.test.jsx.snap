// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Snapshot: Circled 1`] = `
<div
  key="2"
>
  <Icon
    circle={true}
    key="xxs"
    shape="cog"
    size="xxs"
  />
  <Icon
    circle={true}
    key="xs"
    shape="cog"
    size="xs"
  />
  <Icon
    circle={true}
    key="s"
    shape="cog"
    size="s"
  />
  <Icon
    circle={true}
    key="m"
    shape="cog"
    size="m"
  />
  <Icon
    circle={true}
    key="l"
    shape="cog"
    size="l"
  />
  <Icon
    circle={true}
    key="xl"
    shape="cog"
    size="xl"
  />
  <Icon
    circle={true}
    key="xxl"
    shape="cog"
    size="xxl"
  />
</div>
`;

exports[`Icon Snapshot: Circled color 1`] = `
<div
  key="3"
>
  <Icon
    circle={true}
    color="blue"
    key="xxs"
    shape="cog"
    size="xxs"
  />
  <Icon
    circle={true}
    color="blue"
    key="xs"
    shape="cog"
    size="xs"
  />
  <Icon
    circle={true}
    color="blue"
    key="s"
    shape="cog"
    size="s"
  />
  <Icon
    circle={true}
    color="blue"
    key="m"
    shape="cog"
    size="m"
  />
  <Icon
    circle={true}
    color="blue"
    key="l"
    shape="cog"
    size="l"
  />
  <Icon
    circle={true}
    color="blue"
    key="xl"
    shape="cog"
    size="xl"
  />
  <Icon
    circle={true}
    color="blue"
    key="xxl"
    shape="cog"
    size="xxl"
  />
</div>
`;

exports[`Icon Snapshot: Regular 1`] = `
<div
  key="1"
>
  <Icon
    key="xxs"
    shape="cog"
    size="xxs"
  />
  <Icon
    key="xs"
    shape="cog"
    size="xs"
  />
  <Icon
    key="s"
    shape="cog"
    size="s"
  />
  <Icon
    key="m"
    shape="cog"
    size="m"
  />
  <Icon
    key="l"
    shape="cog"
    size="l"
  />
  <Icon
    key="xl"
    shape="cog"
    size="xl"
  />
  <Icon
    key="xxl"
    shape="cog"
    size="xxl"
  />
</div>
`;
